File: src/sections/system.js
const { execFile } = require('child_process');
const os = require('os');
const fs = require('fs');

function run(cmd, args, opts = {}) {
  return new Promise((resolve) => {
    execFile(cmd, args, { timeout: 5000, ...opts }, (err, stdout) => {
      if (err) resolve({ error: err.message, stdout: '' });
      else resolve({ stdout: stdout.trim() });
    });
  });
}

async function getSystem() {
  const info = {};

  // Uptime
  const uptimeSec = os.uptime();
  const days = Math.floor(uptimeSec / 86400);
  const hours = Math.floor((uptimeSec % 86400) / 3600);
  const mins = Math.floor((uptimeSec % 3600) / 60);
  info.uptime = days > 0 ? `${days}d ${hours}h ${mins}m` : `${hours}h ${mins}m`;

  // Load average
  const load = os.loadavg();
  info.load = {
    '1m': load[0].toFixed(2),
    '5m': load[1].toFixed(2),
    '15m': load[2].toFixed(2),
  };
  info.cpus = os.cpus().length;

  // Memory
  const totalMem = os.totalmem();
  const freeMem = os.freemem();
  const usedMem = totalMem - freeMem;
  info.memory = {
    total_gb: (totalMem / 1e9).toFixed(1),
    used_gb: (usedMem / 1e9).toFixed(1),
    free_gb: (freeMem / 1e9).toFixed(1),
    percent: ((usedMem / totalMem) * 100).toFixed(0),
  };

  // Disk usage
  const dfResult = await run('df', ['-h', '--output=target,size,used,avail,pcent', '-x', 'tmpfs', '-x', 'devtmpfs', '-x', 'overlay']);
  if (!dfResult.error && dfResult.stdout) {
    const lines = dfResult.stdout.split('\n').slice(1); // skip header
    info.disks = lines.filter(l => l.trim()).map(line => {
      const parts = line.trim().split(/\s+/);
      const pct = parseInt(parts[4]) || 0;
      return {
        mount: parts[0],
        size: parts[1],
        used: parts[2],
        avail: parts[3],
        percent: pct,
        warning: pct > 80,
      };
    });
    info.diskWarnings = info.disks.filter(d => d.warning).length;
  }

  return info;
}

module.exports = { getSystem };

File: daily_briefing/sections/system.py
"""System health section — uptime, load, memory, disk usage."""

import os
import subprocess
import shutil


def _run(cmd, timeout=5):
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
        return result.stdout.strip(), None
    except (subprocess.TimeoutExpired, FileNotFoundError) as e:
        return "", str(e)


def get_system(args):
    info = {}

    # Uptime
    try:
        with open("/proc/uptime") as f:
            uptime_sec = float(f.read().split()[0])
        days = int(uptime_sec // 86400)
        hours = int((uptime_sec % 86400) // 3600)
        mins = int((uptime_sec % 3600) // 60)
        info["uptime"] = f"{days}d {hours}h {mins}m" if days > 0 else f"{hours}h {mins}m"
    except OSError:
        info["uptime"] = "unknown"

    # Load average
    try:
        load = os.getloadavg()
        info["load"] = {
            "1m": f"{load[0]:.2f}",
            "5m": f"{load[1]:.2f}",
            "15m": f"{load[2]:.2f}",
        }
    except OSError:
        info["load"] = {"1m": "?", "5m": "?", "15m": "?"}

    # CPU count
    info["cpus"] = os.cpu_count() or 0

    # Memory
    try:
        with open("/proc/meminfo") as f:
            meminfo = {}
            for line in f:
                parts = line.split(":")
                if len(parts) == 2:
                    key = parts[0].strip()
                    val = parts[1].strip().split()[0]  # value in kB
                    meminfo[key] = int(val) * 1024  # convert to bytes

        total = meminfo.get("MemTotal", 0)
        available = meminfo.get("MemAvailable", 0)
        used = total - available
        pct = int((used / total) * 100) if total else 0

        info["memory"] = {
            "total_gb": f"{total / 1e9:.1f}",
            "used_gb": f"{used / 1e9:.1f}",
            "free_gb": f"{available / 1e9:.1f}",
            "percent": str(pct),
        }
    except OSError:
        info["memory"] = {"total_gb": "?", "used_gb": "?", "free_gb": "?", "percent": "?"}

    # Disk usage
    out, err = _run(
        ["df", "-h", "--output=target,size,used,avail,pcent", "-x", "tmpfs", "-x", "devtmpfs", "-x", "overlay"]
    )
    if not err and out:
        lines = out.splitlines()[1:]  # skip header
        disks = []
        for line in lines:
            if not line.strip():
                continue
            parts = line.split()
            if len(parts) >= 5:
                pct = int(parts[4].rstrip("%")) if parts[4].rstrip("%").isdigit() else 0
                disks.append({
                    "mount": parts[0],
                    "size": parts[1],
                    "used": parts[2],
                    "avail": parts[3],
                    "percent": pct,
                    "warning": pct > 80,
                })
        info["disks"] = disks
        info["diskWarnings"] = len([d for d in disks if d["warning"]])

    return info
